In [109]:
# imports
import pandas as pd
import numpy as np
import requests

In [4]:
bixi_stations_df = pd.read_csv('/Users/zachrosenthal/Desktop/Data Science/PYTHON/KATAS/Jupyter Notebooks/Statistical-Modelling-Project/data/bixi_stations_df.csv', index_col=0)
bixi_stations_df.head()
# bixi_stations_df.shape

empty_slots  free_bikes                                id   latitude  \
0           12           3  36c6491aa1b52e5ef7005f984738de27  45.516926   
1           13           9  65f5707d61306518123222b4680bc282  45.553219   
2           12          17  0e58810dea9e4869279089cb8576642f  45.526890   
3            9           6  eeda0f4010f34ca2199f3585a76fef66  45.553400   
4           15          19  8140fc3c3486abf913b0bd9211926f2b  45.559842   

   longitude                                               name  
0 -73.564257      Gare d'autocars de Montréal (Berri / Ontario)  
1 -73.539782                                 Marché Maisonneuve  
2 -73.572640                                Rachel / de Brébeuf  
3 -73.662255      Bibliothèque d'Ahuntsic (Lajeunesse / Fleury)  
4 -73.615447  Cité des Arts du Cirque (Paul Boutet / des Reg...

In [3]:
bixi_stations_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 160 entries, 0 to 159
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   empty_slots  160 non-null    int64  
 1   free_bikes   160 non-null    int64  
 2   id           160 non-null    object 
 3   latitude     160 non-null    float64
 4   longitude    160 non-null    float64
 5   name         160 non-null    object 
dtypes: float64(2), int64(2), object(2)
memory usage: 8.8+ KB


In [4]:
bixi_stations_df['latitude'] = bixi_stations_df['latitude'].astype(str)
bixi_stations_df['longitude'] = bixi_stations_df['longitude'].astype(str)
bixi_stations_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 160 entries, 0 to 159
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   empty_slots  160 non-null    int64 
 1   free_bikes   160 non-null    int64 
 2   id           160 non-null    object
 3   latitude     160 non-null    object
 4   longitude    160 non-null    object
 5   name         160 non-null    object
dtypes: int64(2), object(4)
memory usage: 8.8+ KB


# Foursquare

Send a request to Foursquare with a small radius (1000m) for all the bike stations in your city of choice. 

In [123]:
# Started by exploring the Foursquare API documentation. 
#There was also some useful details on how to code the API in Python on their website
#Batched the API calls to reduce the number of requests to the API

# Foursquare API

import requests
import os

def make_api_request(ll, radius, categories, fields):

    url = "https://api.foursquare.com/v3/places/search" # ll={ll}&radius={radius}&categories={categories}&sort=DISTANCE"

    headers = {"accept": "application/json",
               "Authorization": os.environ['MYKEYVARIABLE']}

    params = {
        'll': ll,
        'radius': radius,
        'categories': categories,
        'fields': fields
          }

    response = requests.get(url, params=params, headers=headers)

    return response.json()

#Make sure to append the Lat and lon of the bike stattion to the bar





In [124]:
#shows me the likely API output with the parsed fields I want to use
test_df = pd.DataFrame()
for _, row in bixi_stations_df.head(15).iterrows(): #using a subset of the data to test
    ll = f"{row['latitude']},{row['longitude']}" # make it a string in main verison
    results = make_api_request(ll, 1000, 13003, 'fsq_id,distance,name,geocodes,location,rating,price,popularity')
    test_df = pd.json_normalize(results['results'])

test_df
      
    

fsq_id  distance  \
0  5786aca5498e11e46ccfe8e7       504   
1  4ad4c06ef964a52010fb20e3       318   
2  55aedbd9498eb6db5d095733       434   
3  4b0c1ff8f964a520603723e3       356   
4  56ec4544498e27e294330c2b       715   
5  5536df62498e2325f5f9be5c       395   
6  4c1009053ce120a19f6d071c       265   
7  4ad4c06bf964a5203ef920e3       462   
8  5372ed9f498e6f504d4fe05d       819   
9  4ad4c06af964a5200ff920e3       442   

                                       name  popularity  price  rating  \
0                                    Renard    0.982352    3.0     8.9   
1                              Cabaret Mado    0.963198    2.0     8.3   
2                                 Le 4e Mur    0.966045    3.0     8.5   
3                             Le Saint Bock    0.980111    2.0     8.3   
4                                     Cadet    0.981131    NaN     9.5   
5                              Le Red Tiger    0.968693    2.0     8.2   
6                           Pub l'Ile Noire    0.975758    2.0     7.8   
7  Brasserie Artisanale l'Amère a Boire Inc    0.960864    2.0     8.1   
8                     Bar le Mal Nécessaire    0.981667    3.0     8.8   
9                          Sainte-Elisabeth    0.972632    2.0     8.0   

   geocodes.main.latitude  geocodes.main.longitude  geocodes.roof.latitude  \
0               45.519093               -73.555341               45.519093   
1               45.517753               -73.556994               45.517753   
2               45.516216               -73.565568               45.516216   
3               45.515893               -73.564685               45.515893   
4               45.510388               -73.564365               45.510388   
5               45.519213               -73.557815               45.519213   
6               45.515237               -73.563405               45.515237   
7               45.516435               -73.565962               45.516435   
8               45.508519               -73.561303               45.508519   
9               45.512194               -73.562146               45.512194   

   geocodes.roof.longitude              location.address location.country  \
0               -73.555341   1272 Sainte-Catherine Rue E               CA   
1               -73.556994   1115 Sainte-Catherine Rue E               CA   
2               -73.565568          2021 Saint-Denis Rue               CA   
3               -73.564685          1749 Saint-Denis Rue               CA   
4               -73.564365       1431 Saint-Laurent Blvd               CA   
5               -73.557815  1201 Blvd de Maisonneuve Est               CA   
6               -73.563405          1649 Saint-Denis Rue               CA   
7               -73.565962          2049 Saint-Denis Rue               CA   
8               -73.561303      1106B Saint Laurent Blvd               CA   
9               -73.562146     1412 Sainte-Elisabeth Rue               CA   

                  location.cross_street  \
0                                         
1                            coin Wolfe   
2                           rue Ontario   
3                          coin Ontario   
4  entre De Maisonneuve & Ste-Catherine   
5                                         
6                 entre Emery & Ontario   
7            entre Ontario & Sherbrooke   
8                    coin René-Lévesque   
9                coin Ste-Catherine Est   

                          location.formatted_address location.locality  \
0   1272 Sainte-Catherine Rue E, Montréal QC H2L 2H2          Montréal   
1  1115 Sainte-Catherine Rue E (coin Wolfe), Mont...          Montréal   
2  2021 Saint-Denis Rue (rue Ontario), Montréal Q...          Montréal   
3  1749 Saint-Denis Rue (coin Ontario), Montréal ...          Montréal   
4  1431 Saint-Laurent Blvd (entre De Maisonneuve ...          Montréal   
5  1201 Blvd de Maisonneuve Est, Montréal QC H2L 2A1          Montréal   
6  1649 Saint-Denis Rue (entre Emery & Ontario), ...   

Parse through the response to get the POI (such as restaurants, bars, etc) details you want (ratings, name, location, etc)

In [125]:
# Initialize an empty DataFrame to hold all results
foursquare_bar_df = pd.DataFrame()

for _, row in bixi_stations_df.iterrows():
    ll = f"{row['latitude']},{row['longitude']}"
    # Make the single API request for each latitude and longitude
    api_response = make_api_request(ll, 1000, 13003, 'fsq_id,distance,name,geocodes,location,rating,price,popularity')

    # Normalize the JSON response. Since it’s a loop, the current_result_df will always be the most recent iteration
    current_results_df = pd.json_normalize(api_response['results'])

    # Add a column for the bixi_station id to the DataFrame
    current_results_df['station_id'] = row['id']

    # concatenate the results to the full_results_df
    foursquare_bar_df = pd.concat([foursquare_bar_df, current_results_df], ignore_index=True)

foursquare_bar_df.head()


fsq_id  distance  \
0     55aedbd9498eb6db5d095733     132.0   
1     4b0c1ff8f964a520603723e3     132.0   
2     4ad4c06bf964a5203ef920e3     149.0   
3     56ec4544498e27e294330c2b     741.0   
4     4c1009053ce120a19f6d071c     208.0   
...                        ...       ...   
1438  556902a1498e618f7c287e74     523.0   
1439  4ae0e7f9f964a520928321e3     571.0   
1440  56ec4544498e27e294330c2b     819.0   
1441  57fffed138fa0da136f2a8e3     540.0   
1442  4ae4d07df964a520669e21e3     663.0   

                                          name  popularity  price  rating  \
0                                    Le 4e Mur    0.966045    3.0     8.5   
1                                Le Saint Bock    0.980111    2.0     8.3   
2     Brasserie Artisanale l'Amère a Boire Inc    0.960864    2.0     8.1   
3                                        Cadet    0.981131    NaN     9.5   
4                              Pub l'Ile Noire    0.975758    2.0     7.8   
...                                        ...         ...    ...     ...   
1438                                   Soubois    0.984985    NaN     8.5   
1439                                   Pullman    0.969535    2.0     8.6   
1440                                     Cadet    0.981131    NaN     9.5   
1441                                   La Cage    0.995333    1.0     8.4   
1442                                   Benelux    0.989988    1.0     8.5   

      geocodes.main.latitude  geocodes.main.longitude  geocodes.roof.latitude  \
0                  45.516216               -73.565568               45.516216   
1                  45.515893               -73.564685               45.515893   
2                  45.516435               -73.565962               45.516435   
3                  45.510388               -73.564365               45.510388   
4                  45.515237               -73.563405               45.515237   
...                      ...                      ...                     ...   
1438               45.500634               -73.574708               45.500634   
1439               45.508586               -73.572008               45.508586   
1440               45.510388               -73.564365               45.510388   
1441               45.507661               -73.564375               45.507661   
1442               45.509571               -73.571023               45.509571   

      geocodes.roof.longitude  ...  \
0                  -73.565568  ...   
1                  -73.564685  ...   
2                  -73.565962  ...   
3                  -73.564365  ...   
4                  -73.563405  ...   
...                       ...  ...   
1438               -73.574708  ...   
1439               -73.572008  ...   
1440               -73.564365  ...   
1441               -73.564375  ...   
1442               -73.571023  ...   

                             location.formatted_address location.locality  \
0     2021 Saint-Denis Rue (rue Ontario), Montréal Q...          Montréal   
1     1749 Saint-Denis Rue (coin Ontario), Montréal ...          Montréal   
2     2049 Saint-Denis Rue (entre Ontario & Sherbroo...          Montréal   
3     1431 Saint-Laurent Blvd (entre De Maisonneuve ...          Montréal   
4     1649 Saint-Denis Rue (entre Emery & Ontario), ...          Montréal   
...                                                 ...               ...   
1438  1106 Blvd de Maisonneuve Ouest (entre Peel & S...          Montréal   
1439  3424 Ave du Parc (coin Sherbrooke Ouest), Mont...          Montréal   
1440  1431 Saint-Laurent Blvd (entre De Maisonneuve ...          Montréal   
1441        150 St Catherine Rue W, Montréal QC H2X 3Y2          Montréal   
1442  245 Sherbrooke Rue O (coin Jeanne-Mance), Mont...          Montréal   

     location.postcode location.region geocodes.drop_off.latitude  \
0              H2X 3K8              QC                        NaN   
1              H2X 3K4              QC                        NaN   
2       

In [6]:
foursquare_bar_df = pd.read_csv('/Users/zachrosenthal/Desktop/Data Science/PYTHON/KATAS/Jupyter Notebooks/Statistical-Modelling-Project/data/foursquare_bar_df.csv', index_col=0)
foursquare_bar_df.head(10)

fsq_id  distance  \
0  55aedbd9498eb6db5d095733     132.0   
1  4b0c1ff8f964a520603723e3     132.0   
2  4ad4c06bf964a5203ef920e3     149.0   
3  56ec4544498e27e294330c2b     741.0   
4  4c1009053ce120a19f6d071c     208.0   
5  5786aca5498e11e46ccfe8e7     718.0   
6  4ad4c06bf964a5202af920e3     156.0   
7  4ad4c06ef964a52010fb20e3     578.0   
8  5536df62498e2325f5f9be5c     548.0   
9  54767635498ee8c8c4f3ae4e     158.0   

                                       name  popularity  price  rating  \
0                                 Le 4e Mur    0.966045    3.0     8.5   
1                             Le Saint Bock    0.980111    2.0     8.3   
2  Brasserie Artisanale l'Amère a Boire Inc    0.960864    2.0     8.1   
3                                     Cadet    0.981131    NaN     9.5   
4                           Pub l'Ile Noire    0.975758    2.0     7.8   
5                                    Renard    0.982352    3.0     8.9   
6                           Le Cheval Blanc    0.983543    2.0     7.7   
7                              Cabaret Mado    0.963198    2.0     8.3   
8                              Le Red Tiger    0.968693    2.0     8.2   
9                   Patrick'S Pub Irlandais    0.972054    NaN     7.6   

   geocodes.main.latitude  geocodes.main.longitude  geocodes.roof.latitude  \
0               45.516216               -73.565568               45.516216   
1               45.515893               -73.564685               45.515893   
2               45.516435               -73.565962               45.516435   
3               45.510388               -73.564365               45.510388   
4               45.515237               -73.563405               45.515237   
5               45.519093               -73.555341               45.519093   
6               45.518170               -73.564467               45.518170   
7               45.517753               -73.556994               45.517753   
8               45.519213               -73.557815               45.519213   
9               45.515508               -73.564043               45.515508   

   geocodes.roof.longitude  ...  \
0               -73.565568  ...   
1               -73.564685  ...   
2               -73.565962  ...   
3               -73.564365  ...   
4               -73.563405  ...   
5               -73.555341  ...   
6               -73.564467  ...   
7               -73.556994  ...   
8               -73.557815  ...   
9               -73.564043  ...   

                          location.formatted_address location.locality  \
0  2021 Saint-Denis Rue (rue Ontario), Montréal Q...          Montréal   
1  1749 Saint-Denis Rue (coin Ontario), Montréal ...          Montréal   
2  2049 Saint-Denis Rue (entre Ontario & Sherbroo...          Montréal   
3  1431 Saint-Laurent Blvd (entre De Maisonneuve ...          Montréal   
4  1649 Saint-Denis Rue (entre Emery & Ontario), ...          Montréal   
5   1272 Sainte-Catherine Rue E, Montréal QC H2L 2H2          Montréal   
6  809 Ontario Rue E (entre St-Hubert & St-Christ...          Montréal   
7  1115 Sainte-Catherine Rue E (coin Wolfe), Mont...          Montréal   
8  1201 Blvd de Maisonneuve Est, Montréal QC H2L 2A1          Montréal   
9  1707 Saint-Denis Rue (entre Ontario & De Maiso...          Montréal   

  location.postcode location.region geocodes.drop_off.latitude  \
0           H2X 3K8              QC                        NaN   
1           H2X 3K4              QC                        NaN   
2           H2X 3K8              QC                  45.516262   
3           H2X 2S8              QC                  45.510179   
4           H2X 3K4              QC                        NaN   
5           H2L 2H2              QC                        NaN   
6           H2L 1P1              QC                        NaN   
7           H2L 2G6              QC                        NaN   
8           H2L 2A1              QC                  45.519077   
9           H2X 3K4              Q

In [7]:
foursquare_bar_df.shape

(1443, 23)

Put your parsed results into a DataFrame

In [10]:
foursquare_bar_df
foursquare_bar_df.to_csv('foursquare_bar_df.csv')

# Yelp

Send a request to Yelp with a small radius (1000m) for all the bike stations in your city of choice. 

##### Yelp Fusion
The Yelp Fusion API allows you to get the best local content and user reviews from millions of businesses around the world.

You have the ability to search for businesses by keyword, category, location, price level, phone number. Get rich business data, such as name, address, phone number, photos, Yelp rating, price levels and hours of operation.

##### Foursquare uses lat and long as a combined string or those values. Yelp uses lat and long as a number.
 

In [8]:
# change lat and long values back to numbers for Yelp API
bixi_stations_df['latitude'] = bixi_stations_df['latitude'].astype(float)
bixi_stations_df['longitude'] = bixi_stations_df['longitude'].astype(float)
bixi_stations_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 160 entries, 0 to 159
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   empty_slots  160 non-null    int64  
 1   free_bikes   160 non-null    int64  
 2   id           160 non-null    object 
 3   latitude     160 non-null    float64
 4   longitude    160 non-null    float64
 5   name         160 non-null    object 
dtypes: float64(2), int64(2), object(2)
memory usage: 8.8+ KB


In [14]:
import requests
import os

def make_yelp_api_request(latitude, longitude, radius, term, categories, price):
    api_key = os.environ.get('YELP_API_KEY')
    
    url = "https://api.yelp.com/v3/businesses/search?"

    headers = {
         "Authorization": f"Bearer {api_key}",
        "accept": "application/json"
       }

    params = {
        'latitude': latitude,
        'longitude': longitude,
        'radius': radius, #an integer of meters, use '1000'
        'term': term, # string similar to categories, but from testing it, I see not enough places are actually. Trying 'bar'
        'categories': categories, #an array of strings of business categories. use 'bars'
        'price': price # an array of integers from 1-4
            }
    
    response = requests.get(url, params=params, headers=headers)

    return response.json()


In [194]:
# Test if my API request works:
# Example parameters
latitude = 37.7749   # Example latitude (e.g., San Francisco)
longitude = -122.4194  # Example longitude
radius = 1000  # in meters
term = 'bar'
categories = 'bars'  # Example category
price = '1,2,3,4'  # Price range

# Make the API request
try:
    response = make_yelp_api_request(latitude, longitude, radius, term, categories, price)
    print(response)  # Print the response
except Exception as e:
    print(f"Error: {e}")

{'businesses': [{'id': '2uF7FmpR_QAYGpIzpf_Dfw', 'alias': 'linden-room-san-francisco', 'name': 'Linden Room', 'image_url': 'https://s3-media2.fl.yelpcdn.com/bphoto/IagJcm9FJbZnxFA5ULUC4w/o.jpg', 'is_closed': False, 'url': 'https://www.yelp.com/biz/linden-room-san-francisco?adjust_creative=zuwyQOfmfZK7AOsVQh8PaA&utm_campaign=yelp_api_v3&utm_medium=api_v3_business_search&utm_source=zuwyQOfmfZK7AOsVQh8PaA', 'review_count': 111, 'categories': [{'alias': 'cocktailbars', 'title': 'Cocktail Bars'}], 'rating': 4.5, 'coordinates': {'latitude': 37.7765173, 'longitude': -122.4227968}, 'transactions': [], 'price': '$$$', 'location': {'address1': '330 Gough St', 'address2': '', 'address3': None, 'city': 'San Francisco', 'zip_code': '94102', 'country': 'US', 'state': 'CA', 'display_address': ['330 Gough St', 'San Francisco, CA 94102']}, 'phone': '', 'display_phone': '', 'distance': 348.5251197768975}, {'id': 'bIiFlRWypq30j9YSjEDNcg', 'alias': 'dahlia-lounge-san-francisco', 'name': 'Dahlia Lounge', '

In [ ]:
# # A test df to see what the API returns
import pandas as pd
test_yelp_df = pd.DataFrame()
for _, row in bixi_stations_df.head(15).iterrows(): #using a subset of the data to test
    results = make_yelp_api_request(row['latitude'], row['longitude'], 1000, 'bar', 'bars', [1, 2, 3, 4])
    test_yelp_df = pd.json_normalize(results['businesses'])



In [ ]:
expanded_categories_df = pd.DataFrame()

for _, row in test_yelp_df.iterrows():
    row_categories = pd.json_normalize(row['categories'])
     # Add an identifier column to link back to the original row
    row_categories['id'] = row['id']
    expanded_categories_df = pd.concat([row_categories, expanded_categories_df], ignore_index = True)

expanded_categories_df


Parse through the response to get the POI (such as restaurants, bars, etc) details you want (ratings, name, location, etc)

#### YELP API REQUEST: 

In [33]:
import pandas as pd
# Initialize an empty DataFrame to hold all results
yelp_bar_df = pd.DataFrame()
# Main Yelp API request:
for _, row in bixi_stations_df.iterrows(): 
    results = make_yelp_api_request(row['latitude'], row['longitude'], 1000, 'bar', 'bars', [1, 2, 3, 4])
    current_results_df = pd.json_normalize(results['businesses'])
    # Add a column for the bixi_station id to the DataFrame
    current_results_df['station_id'] = row['id'] # row 'id' is station in bixi df
    # concatenate the results to the current_results_df with the yelp_bar_df
    yelp_bar_df = pd.concat([yelp_bar_df, current_results_df], ignore_index=True)
    





In [34]:

yelp_bar_df.head()

id                           alias  \
0  EQCd_CUFrK_gSARL7s01fw  la-distillerie-no-1-montréal-7   
1  RO32EAy8IfnWAm04vCPmzA              le-4e-mur-montréal   
2  hdJlf4YNdDtQA67nXvFeyw             arcade-mtl-montréal   
3  SbFWAF3Rhtb-vvl5WUqg_A                 yermad-montréal   
4  o7MzamFzaLiCFDmfXAmyRA           le-normandie-montréal   

                  name                                          image_url  \
0  La Distillerie No.1  https://s3-media2.fl.yelpcdn.com/bphoto/HEHvlX...   
1            Le 4e Mur  https://s3-media2.fl.yelpcdn.com/bphoto/5FUAUP...   
2           Arcade MTL  https://s3-media3.fl.yelpcdn.com/bphoto/SxiAd7...   
3              Yer'Mad  https://s3-media2.fl.yelpcdn.com/bphoto/oHnmVp...   
4         Le Normandie  https://s3-media1.fl.yelpcdn.com/bphoto/cSCVUA...   

  is_closed                                                url  review_count  \
0     False  https://www.yelp.com/biz/la-distillerie-no-1-m...         315.0   
1     False  https://www.yelp.com/biz/le-4e-mur-montr%C3%A9...          99.0   
2     False  https://www.yelp.com/biz/arcade-mtl-montr%C3%A...          26.0   
3     False  https://www.yelp.com/biz/yermad-montr%C3%A9al?...          15.0   
4     False  https://www.yelp.com/biz/le-normandie-montr%C3...           8.0   

                                          categories  rating transactions  \
0  [{'alias': 'cocktailbars', 'title': 'Cocktail ...     4.5           []   
1  [{'alias': 'speakeasies', 'title': 'Speakeasie...     4.5           []   
2  [{'alias': 'bars', 'title': 'Bars'}, {'alias':...     4.5           []   
3      [{'alias': 'divebars', 'title': 'Dive Bars'}]     4.5           []   
4  [{'alias': 'bars', 'title': 'Bars'}, {'alias':...     4.5           []   

   ... coordinates.longitude               location.address1  \
0  ...            -73.565401               300 Rue Ontario E   
1  ...            -73.565782            2021 Rue Saint-Denis   
2  ...            -73.565810            2031 Rue Saint-Denis   
3  ...            -73.559534  901 Boulevard de Maisonneuve E   
4  ...            -73.556490                1295 Rue Amherst   

  location.address2  location.address3  location.city  location.zip_code  \
0                                            Montreal            H2X 1H6   
1                                            Montreal            H2X 3K8   
2                                            Montreal            H2X 3K8   
3                                            Montreal            H2L 1Z2   
4                                 None       Montreal            H2L 3K9   

  location.country location.state  \
0               CA             QC   
1               CA             QC   
2               CA             QC   
3               CA             QC   
4               CA             QC   

                            location.display_address  \
0  [300 Rue Ontario E, Montreal, QC H2X 1H6, Canada]   
1  [2021 Rue Saint-Denis, Montreal, QC H2X 3K8, C...   
2  [2031 Rue Saint-Denis, Montreal, QC H2X 3K8, C...   
3  [901 Boulevard de Maisonneuve E, Montreal, QC ...   
4   [1295 Rue Amherst, Montreal, QC H2L 3K9, Canada]   

                         station_id  
0  36c6491aa1b52e5ef7005f984738de27  
1  36c6491aa1b52e5ef7005f984738de27  
2  36c6491aa1b52e5ef7005f984738de27  
3  36c6491aa1b52e5ef7005f984738de27  
4  36c6491aa1b52e5ef7005f984738de27  

[5 rows x 25 columns]

In [35]:
flattened_categories_df = pd.DataFrame()
for _, row in yelp_bar_df.iterrows():
    categories_row_df = pd.json_normalize(row['categories'])
     # Add an identifier column to link back to the original row
    categories_row_df['id'] = row['station_id']
    flattened_categories_df = pd.concat([categories_row_df, flattened_categories_df], ignore_index = True)

flattened_categories_df

#This shows that there is no important information hidden in the categories column. It can be dropped.

#Dropping the 'Categories' Column
yelp_bar_df = yelp_bar_df.drop(columns='categories')


In [36]:
yelp_bar_df.to_csv('yelp_bar_df.csv')

### Put your parsed results into a DataFrame

In [37]:
# Dropping other unnecesary columns
c_yelp_bar_df = yelp_bar_df.copy()
c_yelp_bar_df.keys()

columns_to_drop = ['alias', 'image_url', 'is_closed', 'url', 'transactions',
                   'phone', 'display_phone','location.address2', 'location.address3',
                    'location.city', 'location.zip_code', 'location.country',
                    'location.state', 'location.display_address']
c_yelp_bar_df.drop(columns=columns_to_drop, inplace=True)
c_yelp_bar_df

id                 name  review_count  rating price  \
0     EQCd_CUFrK_gSARL7s01fw  La Distillerie No.1         315.0     4.5    $$   
1     RO32EAy8IfnWAm04vCPmzA            Le 4e Mur          99.0     4.5    $$   
2     hdJlf4YNdDtQA67nXvFeyw           Arcade MTL          26.0     4.5     $   
3     SbFWAF3Rhtb-vvl5WUqg_A              Yer'Mad          15.0     4.5     $   
4     o7MzamFzaLiCFDmfXAmyRA         Le Normandie           8.0     4.5     $   
...                      ...                  ...           ...     ...   ...   
2185  gYAM9MZ_s-YKIUZSZ2Evyg      Les 3 Brasseurs         329.0     3.0    $$   
2186  ol-aOSZbIh46Uy3tKjciaA       Mad Hatter Pub         101.0     4.0     $   
2187  NJDu9wztk9hHyBoBU2TBjg              Benelux         127.0     4.0    $$   
2188  qVdVY4YN0Rr0uKPkE1ltgQ    Le Mal Nécessaire         165.0     4.0    $$   
2189  lLaX4U3_IfiWIFDqv78dcA                Furco          64.0     3.5    $$   

        distance  coordinates.latitude  coordinates.longitude  \
0     250.371652             45.514822             -73.565401   
1     145.030626             45.516178             -73.565782   
2     137.976501             45.516390             -73.565810   
3     369.233928             45.517196             -73.559534   
4     606.689995             45.517311             -73.556490   
...          ...                   ...                    ...   
2185  201.390300             45.502125             -73.570661   
2186  901.414205             45.496716             -73.575265   
2187  663.827615             45.509608             -73.571035   
2188  836.192012             45.508575             -73.561270   
2189  251.885541             45.506480             -73.568690   

                   location.address1                        station_id  
0                  300 Rue Ontario E  36c6491aa1b52e5ef7005f984738de27  
1               2021 Rue Saint-Denis  36c6491aa1b52e5ef7005f984738de27  
2               2031 Rue Saint-Denis  36c6491aa1b52e5ef7005f984738de27  
3     901 Boulevard de Maisonneuve E  36c6491aa1b52e5ef7005f984738de27  
4                   1295 Rue Amherst  36c6491aa1b52e5ef7005f984738de27  
...                              ...                               ...  
2185      732 Rue Sainte-Catherine O  0c1e06741c10b812a0cef5265fa4f76f  
2186               1240 Rue Crescent  0c1e06741c10b812a0cef5265fa4f76f  
2187        245 Rue Sherbrooke Ouest  0c1e06741c10b812a0cef5265fa4f76f  
2188   1106B Saint Laurent Boulevard  0c1e06741c10b812a0cef5265fa4f76f  
2189                   425 Rue Mayor  0c1e06741c10b812a0cef5265fa4f76f  

[2190 rows x 10 columns]

In [38]:
c_yelp_bar_df.shape

(2190, 10)

# Comparing Results

Which API provided you with more complete data? Provide an explanation. 

Yelp's API has better coverage insofar as it was able to find far more bars given the same range and geographic coordinates than Foursquare's API. Also given Yelp's more numerous businesses, it enabled a more precisely filtered API request in the initial request than Foursquare. In my opnion, Yelp has the better coverage as far as getting access to core information regarding the basic data of: location, address, name, etc.

Foursquare, on the other hand, I found had better access and ease of use for data that is beyond the core. It was far easier, through exploring the documentation, to access data like price, popularity, rating, etc via Foursquare than Yelp. Though I will say, Yelp put ratings in their core data where as Foursquare had it in their rich data grouping. Though Foursquare was easier to access in general by using the fields paramter, Yelp had the ratings more generally accessible. Because I am tailoring my exploration towards a ratings analysis to bike usage, Foursquare's ease of use was more beneficial for mu needs. I would like to explore popularity compared to bike usage, but Yelp does not even have a metric for popularity. Additionally, Yelp's price metric of ($, $$, $$$, $$$$) causes needs for additional coding and cleaning steps that Foursquare's does not. Foursqaure uses (1.0, 2.0, 3.0, 4.0). It's far easier to work with numbers than with symbols.

Another interesting thing is the level of precision and detail for the Foursquare API. It gives you a lot of (at least for my purposes) useless information regarding the sepecific geographic coordinates for the roof and front entrance, and other things like that. I'm its practical in some sense, but it seems onerly specific at least to my early/untrained eye.

Something I noticed as well is that even before taking out the duplicates (meaning multiple stations picking up the same bars), Yelp's API managed to pick up 2188 locations given the same bixi stations and radius paramters compared to 1442. While this a simple eye-ball test, it does suggest that Yelp's API has better general coverage. Adding to this position is the fact that the Yelp API request returned 0 nulls once I dropped the columns that were not of interest for my project. Conversly, Foursquare API returned several nulls in the important categories of 'price', 'rating', and 'popularity'. 

Now that I have access to the Yelp API and I have seen this discrepancy between them, it makes me want to add a column in both dataframes to see how many times each bar was picked up by a station. 


#### Get the top 10 Bars according to their rating

### FOURSQUARE

In [213]:
foursquare_bar_df
foursquare_bar_df.isnull().sum()
#I can get rid of columns with 500+ nulls

fsq_id                              0
distance                            0
name                                0
popularity                         35
price                             275
rating                            287
geocodes.main.latitude              0
geocodes.main.longitude             0
geocodes.roof.latitude              9
geocodes.roof.longitude             9
location.address                    5
location.country                    0
location.cross_street             553
location.formatted_address          0
location.locality                   1
location.postcode                  23
location.region                     1
geocodes.drop_off.latitude        566
geocodes.drop_off.longitude       566
station_id                          0
location.address_extended        1342
geocodes.front_door.latitude     1442
geocodes.front_door.longitude    1442
dtype: int64

In [19]:
c_foursquare_bar_df = foursquare_bar_df.copy()
c_foursquare_bar_df.isnull().sum()

fsq_id                              0
distance                            0
name                                0
popularity                         35
price                             275
rating                            287
geocodes.main.latitude              0
geocodes.main.longitude             0
geocodes.roof.latitude              9
geocodes.roof.longitude             9
location.address                    5
location.country                    0
location.cross_street             553
location.formatted_address          0
location.locality                   1
location.postcode                  23
location.region                     1
geocodes.drop_off.latitude        566
geocodes.drop_off.longitude       566
station_id                          0
location.address_extended        1342
geocodes.front_door.latitude     1442
geocodes.front_door.longitude    1442
dtype: int64

In [20]:
c_foursquare_bar_df = foursquare_bar_df.drop(columns=['geocodes.drop_off.latitude', 'geocodes.drop_off.longitude', 'location.address_extended', 'geocodes.front_door.latitude', 'geocodes.front_door.longitude'])
c_oursquare_bar_df = foursquare_bar_df.drop(columns=['location.cross_street', 'geocodes.roof.latitude', 'geocodes.roof.longitude'])

In [21]:
c_foursquare_bar_df.info()


<class 'pandas.core.frame.DataFrame'>
Index: 1443 entries, 0 to 1442
Data columns (total 18 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   fsq_id                      1443 non-null   object 
 1   distance                    1443 non-null   float64
 2   name                        1443 non-null   object 
 3   popularity                  1408 non-null   float64
 4   price                       1168 non-null   float64
 5   rating                      1156 non-null   float64
 6   geocodes.main.latitude      1443 non-null   float64
 7   geocodes.main.longitude     1443 non-null   float64
 8   geocodes.roof.latitude      1434 non-null   float64
 9   geocodes.roof.longitude     1434 non-null   float64
 10  location.address            1438 non-null   object 
 11  location.country            1443 non-null   object 
 12  location.cross_street       890 non-null    object 
 13  location.formatted_address  1443 non-n

In [22]:
c_foursquare_bar_df.isnull().sum()

fsq_id                          0
distance                        0
name                            0
popularity                     35
price                         275
rating                        287
geocodes.main.latitude          0
geocodes.main.longitude         0
geocodes.roof.latitude          9
geocodes.roof.longitude         9
location.address                5
location.country                0
location.cross_street         553
location.formatted_address      0
location.locality               1
location.postcode              23
location.region                 1
station_id                      0
dtype: int64

In [23]:
c_foursquare_bar_df = c_foursquare_bar_df.drop(columns=['geocodes.main.latitude', 'geocodes.main.longitude', 'geocodes.roof.latitude', 'geocodes.roof.longitude'])
# dropped unnecessary geocode columns

c_foursquare_bar_df = c_foursquare_bar_df.drop(columns=['location.address', 'location.country', 'location.cross_street', 'location.formatted_address', 'location.locality', 'location.postcode', 'location.region'])
# dropped unnecessary location columns

In [24]:
print(c_foursquare_bar_df.info())

<class 'pandas.core.frame.DataFrame'>
Index: 1443 entries, 0 to 1442
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   fsq_id      1443 non-null   object 
 1   distance    1443 non-null   float64
 2   name        1443 non-null   object 
 3   popularity  1408 non-null   float64
 4   price       1168 non-null   float64
 5   rating      1156 non-null   float64
 6   station_id  1443 non-null   object 
dtypes: float64(4), object(3)
memory usage: 90.2+ KB
None


In [25]:
c_foursquare_bar_df['name'].value_counts()

name
Escondite                      43
Nacarat                        25
Cadet                          23
Ye Olde Orchard Pub & Grill    22
Nyks Bistro Pub                22
                               ..
La Distillerie                  1
Notre Dame des Quilles          1
Taverne Cobra                   1
La Boîte à Karaoke              1
Les Enfants du Rock MTL         1
Name: count, Length: 300, dtype: int64

#### Cleaning the FOURSQUARE DF, Replacing the NaN Values: 

In [26]:
# FILL IN 'POPULARITY' NULLS:
print('The mean popularity of bars on foursquare for the data considered is:', c_foursquare_bar_df['popularity'].mean())
print('The median popularity of bars on foursquare for the data considered is:', c_foursquare_bar_df['popularity'].median())
#since the popularitu category has only 35 null entries in the column, I will replace the nulls with the mean.

c_foursquare_bar_mean = c_foursquare_bar_df['popularity'].mean()
c_foursquare_bar_df['popularity'] = c_foursquare_bar_df['popularity'].fillna(c_foursquare_bar_mean)
# FILL IN 'PRICE' NULLS:
print('The mean price of bars on foursquare for the data considered is:', c_foursquare_bar_df['price'].mean())
print('The median price of bars on foursquare for the data considered is:', c_foursquare_bar_df['price'].median())
print('The mode price of bars on foursquare for the data considered is:', c_foursquare_bar_df['price'].mode())

# All three central tendencies are close to 2.0 so I will fill NaN with 2.0
c_foursquare_bar_median = c_foursquare_bar_df['price'].median()
c_foursquare_bar_df['price'] = c_foursquare_bar_df['price'].fillna(c_foursquare_bar_median)

#FILL IN 'RATING' NULLS:
print(c_foursquare_bar_df['rating'])
print('The mean rating of bars on foursquare for the data considered is:', c_foursquare_bar_df['rating'].mean())
print('The median rating of bars on foursquare for the data considered is:', c_foursquare_bar_df['rating'].median())
print('The mode rating of bars on foursquare for the data considered is:', c_foursquare_bar_df['rating'].mode())

#Since the median and mode are 8.3 and the mean is 8.17, I'm going to fill the NaN values in the rating column with 8.2
c_foursquare_bar_fill_rating = 8.2
c_foursquare_bar_df['rating'] = c_foursquare_bar_df['rating'].fillna(8.2)

The mean popularity of bars on foursquare for the data considered is: 0.8996190724673447
The median popularity of bars on foursquare for the data considered is: 0.9686926859206292
The mean price of bars on foursquare for the data considered is: 2.041095890410959
The median price of bars on foursquare for the data considered is: 2.0
The mode price of bars on foursquare for the data considered is: 0    2.0
Name: price, dtype: float64
0       8.5
1       8.3
2       8.1
3       9.5
4       7.8
       ... 
1438    8.5
1439    8.6
1440    9.5
1441    8.4
1442    8.5
Name: rating, Length: 1443, dtype: float64
The mean rating of bars on foursquare for the data considered is: 8.173875432525952
The median rating of bars on foursquare for the data considered is: 8.3
The mode rating of bars on foursquare for the data considered is: 0    8.3
Name: rating, dtype: float64


In [27]:
c_foursquare_bar_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1443 entries, 0 to 1442
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   fsq_id      1443 non-null   object 
 1   distance    1443 non-null   float64
 2   name        1443 non-null   object 
 3   popularity  1443 non-null   float64
 4   price       1443 non-null   float64
 5   rating      1443 non-null   float64
 6   station_id  1443 non-null   object 
dtypes: float64(4), object(3)
memory usage: 90.2+ KB


In [28]:
# Create a new column 'total_counts' with the count of each 'name'
c_foursquare_bar_df['total_counts'] = c_foursquare_bar_df.groupby('name')['name'].transform('size')
c_foursquare_bar_df

fsq_id  distance  \
0     55aedbd9498eb6db5d095733     132.0   
1     4b0c1ff8f964a520603723e3     132.0   
2     4ad4c06bf964a5203ef920e3     149.0   
3     56ec4544498e27e294330c2b     741.0   
4     4c1009053ce120a19f6d071c     208.0   
...                        ...       ...   
1438  556902a1498e618f7c287e74     523.0   
1439  4ae0e7f9f964a520928321e3     571.0   
1440  56ec4544498e27e294330c2b     819.0   
1441  57fffed138fa0da136f2a8e3     540.0   
1442  4ae4d07df964a520669e21e3     663.0   

                                          name  popularity  price  rating  \
0                                    Le 4e Mur    0.966045    3.0     8.5   
1                                Le Saint Bock    0.980111    2.0     8.3   
2     Brasserie Artisanale l'Amère a Boire Inc    0.960864    2.0     8.1   
3                                        Cadet    0.981131    2.0     9.5   
4                              Pub l'Ile Noire    0.975758    2.0     7.8   
...                                        ...         ...    ...     ...   
1438                                   Soubois    0.984985    2.0     8.5   
1439                                   Pullman    0.969535    2.0     8.6   
1440                                     Cadet    0.981131    2.0     9.5   
1441                                   La Cage    0.995333    1.0     8.4   
1442                                   Benelux    0.989988    1.0     8.5   

                            station_id  total_counts  
0     36c6491aa1b52e5ef7005f984738de27            11  
1     36c6491aa1b52e5ef7005f984738de27             9  
2     36c6491aa1b52e5ef7005f984738de27             8  
3     36c6491aa1b52e5ef7005f984738de27            23  
4     36c6491aa1b52e5ef7005f984738de27             6  
...                                ...           ...  
1438  0c1e06741c10b812a0cef5265fa4f76f            18  
1439  0c1e06741c10b812a0cef5265fa4f76f            16  
1440  0c1e06741c10b812a0cef5265fa4f76f            23  
1441  0c1e06741c10b812a0cef5265fa4f76f            12  
1442  0c1e06741c10b812a0cef5265fa4f76f            11  

[1443 rows x 8 columns]

In [58]:
c_foursquare_bar_df.to_csv('clean_foursquare_bar_df.csv')
clean_foursquare_bar_df  = pd.read_csv('/Users/zachrosenthal/Desktop/Data Science/PYTHON/KATAS/Jupyter Notebooks/Statistical-Modelling-Project/data/clean_foursquare_bar_df.csv', 
                                       index_col = 0)

### Cleaning the YELP DF and REPLACING NaN Values: 

In [39]:
print(c_yelp_bar_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2190 entries, 0 to 2189
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     2190 non-null   object 
 1   name                   2190 non-null   object 
 2   review_count           2190 non-null   float64
 3   rating                 2190 non-null   float64
 4   price                  2190 non-null   object 
 5   distance               2190 non-null   float64
 6   coordinates.latitude   2190 non-null   float64
 7   coordinates.longitude  2190 non-null   float64
 8   location.address1      2190 non-null   object 
 9   station_id             2190 non-null   object 
dtypes: float64(5), object(5)
memory usage: 171.2+ KB
None


In [40]:
c_yelp_bar_df.isnull().sum()
#well that makes cleaning an easy

id                       0
name                     0
review_count             0
rating                   0
price                    0
distance                 0
coordinates.latitude     0
coordinates.longitude    0
location.address1        0
station_id               0
dtype: int64

In [42]:
# Create a new column 'total_counts' with the count of each 'name'
c_yelp_bar_df['total_counts'] = c_yelp_bar_df.groupby('name')['name'].transform('size')
c_yelp_bar_df.head()

id                 name  review_count  rating price  \
0  EQCd_CUFrK_gSARL7s01fw  La Distillerie No.1         315.0     4.5    $$   
1  RO32EAy8IfnWAm04vCPmzA            Le 4e Mur          99.0     4.5    $$   
2  hdJlf4YNdDtQA67nXvFeyw           Arcade MTL          26.0     4.5     $   
3  SbFWAF3Rhtb-vvl5WUqg_A              Yer'Mad          15.0     4.5     $   
4  o7MzamFzaLiCFDmfXAmyRA         Le Normandie           8.0     4.5     $   

     distance  coordinates.latitude  coordinates.longitude  \
0  250.371652             45.514822             -73.565401   
1  145.030626             45.516178             -73.565782   
2  137.976501             45.516390             -73.565810   
3  369.233928             45.517196             -73.559534   
4  606.689995             45.517311             -73.556490   

                location.address1                        station_id  \
0               300 Rue Ontario E  36c6491aa1b52e5ef7005f984738de27   
1            2021 Rue Saint-Denis  36c6491aa1b52e5ef7005f984738de27   
2            2031 Rue Saint-Denis  36c6491aa1b52e5ef7005f984738de27   
3  901 Boulevard de Maisonneuve E  36c6491aa1b52e5ef7005f984738de27   
4                1295 Rue Amherst  36c6491aa1b52e5ef7005f984738de27   

   total_counts  
0            21  
1            17  
2            13  
3            11  
4            14

In [43]:
print(c_yelp_bar_df.keys())
columns_to_drop = ['distance', 'coordinates.latitude', 'coordinates.longitude', 'location.address1']
c_yelp_bar_df = c_yelp_bar_df.drop(columns=columns_to_drop)

Index(['id', 'name', 'review_count', 'rating', 'price', 'distance',
       'coordinates.latitude', 'coordinates.longitude', 'location.address1',
       'station_id', 'total_counts'],
      dtype='object')


In [44]:
copy_yelp_bar_df = c_yelp_bar_df.copy()


# Usa a lambda function to make the small changes in the price column. 
# If it works, its a good way to convert the rating column to be out of 10 like in foursquare.
copy_yelp_bar_df['price'] = copy_yelp_bar_df['price'].apply(lambda x: len(x))

#convert the price column to a float so it is in the same style as foursquare
copy_yelp_bar_df['price'] = copy_yelp_bar_df['price'].astype(float)

c_yelp_bar_df = copy_yelp_bar_df


In [45]:
# #Checking for how the rows are duplicated:
duplicated_rows_yelp_bar_df = c_yelp_bar_df[c_yelp_bar_df.duplicated(['name', 'id'])]
duplicated_rows_yelp_bar_df #1934 duplicated rows

c_yelp_bar_df.drop_duplicates(subset = ['id', 'name'], keep='first', inplace=True)

In [52]:
clean_yelp_bar_df = pd.read_csv('/Users/zachrosenthal/Desktop/Data Science/PYTHON/KATAS/Jupyter Notebooks/Statistical-Modelling-Project/data/clean_yelp_bar_df.csv', index_col=0)
clean_yelp_bar_df.describe()

review_count      rating       price  total_counts
count    255.000000  255.000000  255.000000    255.000000
mean      63.886275    3.741176    1.956863      8.913725
std       81.178953    0.639628    0.616459      5.922765
min        1.000000    1.000000    1.000000      1.000000
25%       14.000000    3.500000    2.000000      5.000000
50%       36.000000    4.000000    2.000000      7.000000
75%       75.500000    4.000000    2.000000     12.000000
max      598.000000    5.000000    4.000000     29.000000

### The top 10 Montreal bars based on Foursquare's Ratings and Popularity respectively are:

In [47]:
best_rating = c_foursquare_bar_df['rating'].max()
print('The highest rating for any fs_bar within 1000m of a bixi stop is:', best_rating)


count_best_bar_rating = c_foursquare_bar_df[c_foursquare_bar_df['rating'] == best_rating].value_counts().sum()
print('The number of bars with the max rating of 9.5 is:', count_best_bar_rating)

#This shows me that the bike stations all picked up this one bar 'Cadet'. I will have to clean this later.
names_best_rating_fs_bars = c_foursquare_bar_df[c_foursquare_bar_df['rating'] == best_rating]['name'] 
print(names_best_rating_fs_bars)





The highest rating for any fs_bar within 1000m of a bixi stop is: 9.5
The number of bars with the max rating of 9.5 is: 23
3       Cadet
110     Cadet
155     Cadet
185     Cadet
221     Cadet
236     Cadet
244     Cadet
462     Cadet
478     Cadet
556     Cadet
669     Cadet
801     Cadet
823     Cadet
832     Cadet
838     Cadet
877     Cadet
928     Cadet
949     Cadet
981     Cadet
1121    Cadet
1239    Cadet
1404    Cadet
1440    Cadet
Name: name, dtype: object


In [59]:
clean_foursquare_bar_df = clean_foursquare_bar_df.drop_duplicates( subset=['fsq_id', 'bar_name','popularity', 'price', 'rating'], 
                                                              keep='first', 
                                                              inplace=False )
clean_foursquare_bar_df.head()


fsq_id                                  bar_name  \
0  55aedbd9498eb6db5d095733                                 Le 4e Mur   
1  4b0c1ff8f964a520603723e3                             Le Saint Bock   
2  4ad4c06bf964a5203ef920e3  Brasserie Artisanale l'Amère a Boire Inc   
3  56ec4544498e27e294330c2b                                     Cadet   
4  4c1009053ce120a19f6d071c                           Pub l'Ile Noire   

                         station_id  popularity  price  rating  total_counts  
0  36c6491aa1b52e5ef7005f984738de27    0.966045    3.0     8.5            11  
1  36c6491aa1b52e5ef7005f984738de27    0.980111    2.0     8.3             9  
2  36c6491aa1b52e5ef7005f984738de27    0.960864    2.0     8.1             8  
3  36c6491aa1b52e5ef7005f984738de27    0.981131    2.0     9.5            23  
4  36c6491aa1b52e5ef7005f984738de27    0.975758    2.0     7.8             6

### Best RATED Bars in Montreal according to Foursquare:

In [60]:
#Best RATED Bars in Montreal according to Foursquare:
best_rated_fs_bars = clean_foursquare_bar_df.sort_values(['rating'], ascending=False).head(10)
best_rated_fs_bars

fsq_id                bar_name  \
3    56ec4544498e27e294330c2b                   Cadet   
72   58c8b974951e7d7e08bc6fd8                 Darling   
103  540314f8498e97602cef55ba           Isle de Garde   
75   4ad4c06bf964a5203df920e3  Brasserie Dieu du Ciel   
176  51d36b04498e0485ae7335e3         Le Vin Papillon   
71   4b9ae085f964a520a7de35e3     Buvette Chez Simone   
74   533de7ac498e9266dca655ff           Le Majestique   
110  52647dcd498e25e8a0faceb5             MaBrasserie   
58   5849d31423188e1e7a58e397                 Marconi   
130  59622257531593143d4e0ff9                 Nacarat   

                           station_id  popularity  price  rating  total_counts  
3    36c6491aa1b52e5ef7005f984738de27    0.981131    2.0     9.5            23  
72   a32e2261fca6732a8f47eec9fef69017    0.979904    3.0     9.2            10  
103  c8aba35f90d2eeada8a20ea7915c943a    0.970612    3.0     9.1             4  
75   a32e2261fca6732a8f47eec9fef69017    0.962434    2.0     9.1             8  
176  444448176f019fa75930cba6037f6ed8    0.951130    3.0     9.1            11  
71   a32e2261fca6732a8f47eec9fef69017    0.944900    2.0     9.0             6  
74   a32e2261fca6732a8f47eec9fef69017    0.976735    2.0     9.0            10  
110  e335f2f5ad674c200bb9795ee863749c    0.965760    2.0     9.0             5  
58   00c210cb99cf9d1b923c1548938aee56    0.964168    2.0     9.0             5  
130  cecb49c91023c7469470569da814659e    0.982559    3.0     8.9            25

### 10 Best Rated Bars According to Yelp:

In [460]:
best_rated_y_bars = clean_yelp_bar_df.sort_values('rating', ascending=False).head(10)
best_rated_y_bars

id                    name  review_count  rating  \
874  kYWKmXgfcN1nzt50Ulgcyw          Cafe Del Monko           1.0     5.0   
295  evTWdlBsQl6EMiVcm0fXww           Cloakroom Bar         194.0     5.0   
50   xxs6Jni9Fcq7Dw9YArR3ew              Caffè Roma           3.0     5.0   
452  ghwcjmCP3SEGzVPpZE_4Dg                Monopole          21.0     5.0   
88   xM5jI2jE7-R-qhP9CxriwA  Notre Dame des Quilles          36.0     4.5   
115  cLkyosdtcATTRTXX9EmHvg              North Star          30.0     4.5   
113  EpAXhhvS-8U9aG2LlAr5WA           Le Majestique         204.0     4.5   
110  qXmg2Y6c-Le7KAoL4apHfg     Nouveau Cafe Cozmos           4.0     4.5   
511  oq3X84QyoUC8tLDyOJUXRQ                  Else's         122.0     4.5   
96   NG3TF5xwiF76S0N8SZ8xuQ         Bruno Sport Bar          10.0     4.5   

     price                        station_id  total_counts  
874    1.0  d63041391198f1496b84b02d71a3ed90             3  
295    3.0  69f583a85e1a83b443c203b781bfc0c9            22  
50     1.0  8140fc3c3486abf913b0bd9211926f2b             7  
452    2.0  ede6b6e636d0057c3569441aad58f64c            11  
88     1.0  00c210cb99cf9d1b923c1548938aee56             6  
115    2.0  a32e2261fca6732a8f47eec9fef69017            12  
113    2.0  a32e2261fca6732a8f47eec9fef69017            12  
110    2.0  f9bb106633ce38eeec6728faed3155ad             5  
511    2.0  0c9cd68aa1bd69a080463ce7f302337c             9  
96     1.0  00c210cb99cf9d1b923c1548938aee56             6

There is something really interesting going on when the two sites, using the same Bixi station locations actually show different bars in the top 10. The most obvious difference, is in the price. Foursquare's top 10 rated bars skew further towards the more expensive bars than the one from Yelp.

On Yelp, only 1/10 of the best rated bars have a price value of 3.0 whereas  on Foursquare, 4/10 bars in the top 10 have a price value of 3.0. On the other side of the spectrum, on Yelp, 4/10 bars in the 10 best rated group have a price value of  1.0 whereas 0/10 bars on Foursquare in top 10 best-rated bars have a price value of 1.0. This would be an interesting insight for either of those companies to better explore their clientel and tailor their offerings. A good place for future study. 

In [62]:
print('FS DF SHAPE:', clean_foursquare_bar_df.shape)
print('YELP DF SHAPE', clean_yelp_bar_df.shape)
# I need to drop the distance from the FS df
# Getting the shapes so I can see if the column numbers are the same to allow for easier joining

FS DF SHAPE: (304, 7)
YELP DF SHAPE (255, 7)


#### Joining both cleaned bar DFs:

In [66]:
clean_yelp_bar_df = pd.read_csv('/Users/zachrosenthal/Desktop/Data Science/PYTHON/KATAS/Jupyter Notebooks/Statistical-Modelling-Project/data/clean_yelp_bar_df.csv', index_col = 0)
clean_yelp_bar_df.head()

y_id              bar_name  \
0  EQCd_CUFrK_gSARL7s01fw   La Distillerie No.1   
1  RO32EAy8IfnWAm04vCPmzA             Le 4e Mur   
2  XQDWifQPvL2R_c3ptJ8D2g  Randolph Pub Ludique   
3  hdJlf4YNdDtQA67nXvFeyw            Arcade MTL   
4  SbFWAF3Rhtb-vvl5WUqg_A               Yer'Mad   

                         station_id  review_count  rating  price  total_counts  
0  36c6491aa1b52e5ef7005f984738de27         315.0     4.5    2.0            21  
1  36c6491aa1b52e5ef7005f984738de27          99.0     4.5    2.0            17  
2  36c6491aa1b52e5ef7005f984738de27          96.0     4.5    2.0            16  
3  36c6491aa1b52e5ef7005f984738de27          26.0     4.5    1.0            13  
4  36c6491aa1b52e5ef7005f984738de27          15.0     4.5    1.0            11

In [73]:
# Merging the Yelp and Foursquare Dataframes
fsq_y_bar_df = pd.merge(clean_foursquare_bar_df, clean_yelp_bar_df, how='outer', on=['bar_name', 'station_id'])
fsq_y_bar_df.head()
fsq_y_bar_df.to_csv('/Users/zachrosenthal/Desktop/Data Science/PYTHON/KATAS/Jupyter Notebooks/Statistical-Modelling-Project/data/fsq_y_bar_df.csv')


In [144]:
fsq_y_bar_df = pd.read_csv('/Users/zachrosenthal/Desktop/Data Science/PYTHON/KATAS/Jupyter Notebooks/Statistical-Modelling-Project/data/fsq_y_bar_df.csv', index_col=0)
fsq_y_bar_df.head(20)
duplicates = fsq_y_bar_df.duplicated(subset='bar_name')
print('There were this many bars picked up by both API requests:', 
      duplicates.sum()) #39 dublicated bars that were picked up by both API calls

There were this many bars picked up by both API requests: 39


#### The fact that there are only 39 bars that appear in both. This reflects the fact that the two websites only shared 1 bar in the top 10. Also the fact that only 39 bars are shared adds more evidence (including the differences in price) that these two sites whether intentionally or otherwise are catered to different clientel. 

In [145]:
#Dropping the dublicated bars:
fsq_y_bar_df.drop_duplicates(subset='bar_name', keep='first', inplace=True)
print('The shape after duplicates are dropped', fsq_y_bar_df.shape)

The shape after duplicates are dropped (487, 12)


In [80]:
fsq_y_bar_df.head()

fsq_id                                  bar_name  \
0  55aedbd9498eb6db5d095733                                 Le 4e Mur   
1  4b0c1ff8f964a520603723e3                             Le Saint Bock   
2  4ad4c06bf964a5203ef920e3  Brasserie Artisanale l'Amère a Boire Inc   
3  56ec4544498e27e294330c2b                                     Cadet   
4  4c1009053ce120a19f6d071c                           Pub l'Ile Noire   

                         station_id  popularity  price_x  rating_x  \
0  36c6491aa1b52e5ef7005f984738de27    0.966045      3.0       8.5   
1  36c6491aa1b52e5ef7005f984738de27    0.980111      2.0       8.3   
2  36c6491aa1b52e5ef7005f984738de27    0.960864      2.0       8.1   
3  36c6491aa1b52e5ef7005f984738de27    0.981131      2.0       9.5   
4  36c6491aa1b52e5ef7005f984738de27    0.975758      2.0       7.8   

   total_counts_x                    y_id  review_count  rating_y  price_y  \
0            11.0  RO32EAy8IfnWAm04vCPmzA          99.0       4.5      2.0   
1             9.0                     NaN           NaN       NaN      NaN   
2             8.0                     NaN           NaN       NaN      NaN   
3            23.0                     NaN           NaN       NaN      NaN   
4             6.0                     NaN           NaN       NaN      NaN   

   total_counts_y  
0            17.0  
1             NaN  
2             NaN  
3             NaN  
4             NaN

#### Filling out the NaN values after joining the two dataframes:

In [146]:
# Making a new price column that combines the price data from both.
# They are both on the same scale (1.0, 2.0, 3.0, 4.0)
# price_x is fsq, price_y is yelp
fsq_y_bar_df['price'] = fsq_y_bar_df.apply(lambda row: row['price_y'] if pd.isna(row['price_x']) else row['price_x'], axis=1)
fsq_y_bar_df.drop(columns=['price_x', 'price_y'], inplace=True)
fsq_y_bar_df.head()


fsq_id                                  bar_name  \
0  55aedbd9498eb6db5d095733                                 Le 4e Mur   
1  4b0c1ff8f964a520603723e3                             Le Saint Bock   
2  4ad4c06bf964a5203ef920e3  Brasserie Artisanale l'Amère a Boire Inc   
3  56ec4544498e27e294330c2b                                     Cadet   
4  4c1009053ce120a19f6d071c                           Pub l'Ile Noire   

                         station_id  popularity  rating_x  total_counts_x  \
0  36c6491aa1b52e5ef7005f984738de27    0.966045       8.5            11.0   
1  36c6491aa1b52e5ef7005f984738de27    0.980111       8.3             9.0   
2  36c6491aa1b52e5ef7005f984738de27    0.960864       8.1             8.0   
3  36c6491aa1b52e5ef7005f984738de27    0.981131       9.5            23.0   
4  36c6491aa1b52e5ef7005f984738de27    0.975758       7.8             6.0   

                     y_id  review_count  rating_y  total_counts_y  price  
0  RO32EAy8IfnWAm04vCPmzA          99.0       4.5            17.0    3.0  
1                     NaN           NaN       NaN             NaN    2.0  
2                     NaN           NaN       NaN             NaN    2.0  
3                     NaN           NaN       NaN             NaN    2.0  
4                     NaN           NaN       NaN             NaN    2.0

In [147]:
#total_counts are also on the same scale for yelp and fsq so those columns can be combined
# total_counts_x is fsq, total_counts_y is yelp
fsq_y_bar_df['total_counts'] = fsq_y_bar_df.apply(lambda row: row['total_counts_y'] if pd.isna(row['total_counts_x']) else row['total_counts_x'], axis=1)
fsq_y_bar_df.drop(columns=['total_counts_x', 'total_counts_y'], inplace=True) #dropping the original total_counts columns
fsq_y_bar_df.head()

fsq_id                                  bar_name  \
0  55aedbd9498eb6db5d095733                                 Le 4e Mur   
1  4b0c1ff8f964a520603723e3                             Le Saint Bock   
2  4ad4c06bf964a5203ef920e3  Brasserie Artisanale l'Amère a Boire Inc   
3  56ec4544498e27e294330c2b                                     Cadet   
4  4c1009053ce120a19f6d071c                           Pub l'Ile Noire   

                         station_id  popularity  rating_x  \
0  36c6491aa1b52e5ef7005f984738de27    0.966045       8.5   
1  36c6491aa1b52e5ef7005f984738de27    0.980111       8.3   
2  36c6491aa1b52e5ef7005f984738de27    0.960864       8.1   
3  36c6491aa1b52e5ef7005f984738de27    0.981131       9.5   
4  36c6491aa1b52e5ef7005f984738de27    0.975758       7.8   

                     y_id  review_count  rating_y  price  total_counts  
0  RO32EAy8IfnWAm04vCPmzA          99.0       4.5    3.0          11.0  
1                     NaN           NaN       NaN    2.0           9.0  
2                     NaN           NaN       NaN    2.0           8.0  
3                     NaN           NaN       NaN    2.0          23.0  
4                     NaN           NaN       NaN    2.0           6.0

In [155]:
# Standardizing the Rating columns
# I'll  multiply the Yelp rating by 2 to put it on a 10-scale to match Foursquares rating metric.

# Standardizing the Yelp rating column: 
# Multiply the yelp_rating column by 2
fsq_y_bar_df['rating_y'].value_counts() #Checking to see thw counts before
fsq_y_bar_df['rating_y'] = fsq_y_bar_df.apply(lambda row: ((row['rating_y'])*2), axis=1)
fsq_y_bar_df['rating_y'].value_counts() #Sanity check to see it worked properly
fsq_y_bar_df.head() 
# The first rating is invalid, but its ok, when combining the two ratings columns, the fsq rating will be prioritized.

fsq_y_bar_df['rating_x'].value_counts()
# Combining the rating_x and rating_y columns into one 'rating' column: 
fsq_y_bar_df['rating'] = fsq_y_bar_df.apply(lambda row: row['rating_x'] if pd.isna(row['rating_y']) else row['rating_y'], axis=1)
# in the new rating column, use rating_x if rating_y is null, otherwise use rating_y 
fsq_y_bar_df['rating'].value_counts()

fsq_y_bar_df.drop(columns=['rating_x', 'rating_y'], inplace=True) #dropping the original rating columns
fsq_y_bar_df.head()

fsq_id                                  bar_name  \
0  55aedbd9498eb6db5d095733                                 Le 4e Mur   
1  4b0c1ff8f964a520603723e3                             Le Saint Bock   
2  4ad4c06bf964a5203ef920e3  Brasserie Artisanale l'Amère a Boire Inc   
3  56ec4544498e27e294330c2b                                     Cadet   
4  4c1009053ce120a19f6d071c                           Pub l'Ile Noire   

                         station_id  popularity                    y_id  \
0  36c6491aa1b52e5ef7005f984738de27    0.966045  RO32EAy8IfnWAm04vCPmzA   
1  36c6491aa1b52e5ef7005f984738de27    0.980111                     NaN   
2  36c6491aa1b52e5ef7005f984738de27    0.960864                     NaN   
3  36c6491aa1b52e5ef7005f984738de27    0.981131                     NaN   
4  36c6491aa1b52e5ef7005f984738de27    0.975758                     NaN   

   review_count  price  total_counts  rating  
0          99.0    3.0          11.0     9.0  
1           NaN    2.0           9.0     8.3  
2           NaN    2.0           8.0     8.1  
3           NaN    2.0          23.0     9.5  
4           NaN    2.0           6.0     7.8

In [156]:
invalid_bar_rating = fsq_y_bar_df.loc[(fsq_y_bar_df['rating'] > 10)]
invalid_bar_rating
# There are no records with invalid ratings which are above 10.0. 




Empty DataFrame
Columns: [fsq_id, bar_name, station_id, popularity, y_id, review_count, price, total_counts, rating]
Index: []

#### Now the tricky part is to fill in the NaNs for the Popularity and review_count since yelp did not have a popularity rating and foursquare didn't have a review_count columns.



In [157]:
print(fsq_y_bar_df.describe())



       popularity  review_count       price  total_counts      rating
count  300.000000    219.000000  487.000000    487.000000  487.000000
mean     0.838696     60.000000    1.977413      6.236140    7.722382
std      0.257324     78.933952    0.592740      5.562189    1.033399
min      0.019968      1.000000    1.000000      1.000000    2.000000
25%      0.899619     12.500000    2.000000      2.000000    7.000000
50%      0.954887     35.000000    2.000000      4.000000    8.000000
75%      0.974516     68.000000    2.000000      8.000000    8.200000
max      0.998544    598.000000    4.000000     43.000000   10.000000


In [158]:
fsq_y_bar_df.isnull().sum()

fsq_id          187
bar_name          0
station_id        0
popularity      187
y_id            268
review_count    268
price             0
total_counts      0
rating            0
dtype: int64

In [160]:
# fsq_y_bar_df.to_csv('/Users/zachrosenthal/Desktop/Data Science/PYTHON/KATAS/Jupyter Notebooks/Statistical-Modelling-Project/data/fsq_y_bar_nulls_in_popularity_and_reviews_df.csv')

#Incase the random filling of NaNs goes badly, use this to go back to the nearest safe point:
fsq_y_bar_with_nulls_df = pd.read_csv('/Users/zachrosenthal/Desktop/Data Science/PYTHON/KATAS/Jupyter Notebooks/Statistical-Modelling-Project/data/fsq_y_bar_nulls_in_popularity_and_reviews_df.csv', index_col=0)
fsq_y_bar_with_nulls_df.head()

fsq_id                                  bar_name  \
0  55aedbd9498eb6db5d095733                                 Le 4e Mur   
1  4b0c1ff8f964a520603723e3                             Le Saint Bock   
2  4ad4c06bf964a5203ef920e3  Brasserie Artisanale l'Amère a Boire Inc   
3  56ec4544498e27e294330c2b                                     Cadet   
4  4c1009053ce120a19f6d071c                           Pub l'Ile Noire   

                         station_id  popularity                    y_id  \
0  36c6491aa1b52e5ef7005f984738de27    0.966045  RO32EAy8IfnWAm04vCPmzA   
1  36c6491aa1b52e5ef7005f984738de27    0.980111                     NaN   
2  36c6491aa1b52e5ef7005f984738de27    0.960864                     NaN   
3  36c6491aa1b52e5ef7005f984738de27    0.981131                     NaN   
4  36c6491aa1b52e5ef7005f984738de27    0.975758                     NaN   

   review_count  price  total_counts  rating  
0          99.0    3.0          11.0     9.0  
1           NaN    2.0           9.0     8.3  
2           NaN    2.0           8.0     8.1  
3           NaN    2.0          23.0     9.5  
4           NaN    2.0           6.0     7.8

In [161]:
fsq_y_bar_with_nulls_df['popularity'].describe()

count    300.000000
mean       0.838696
std        0.257324
min        0.019968
25%        0.899619
50%        0.954887
75%        0.974516
max        0.998544
Name: popularity, dtype: float64

In [162]:

#Filling the missing popularity values with random numbers within a uniform distribution:
fsq_y_bar_with_nulls_df['popularity'] = fsq_y_bar_with_nulls_df.apply(lambda row: np.random.rand() if pd.isna(row['popularity']) else row['popularity'], axis=1)
fsq_y_bar_with_nulls_df['popularity'].describe()


count    487.000000
mean       0.718477
std        0.312244
min        0.003384
25%        0.476300
50%        0.899619
75%        0.965410
max        0.998544
Name: popularity, dtype: float64

In [163]:
fsq_y_bar_with_nulls_df.isnull().sum()

fsq_id          187
bar_name          0
station_id        0
popularity        0
y_id            268
review_count    268
price             0
total_counts      0
rating            0
dtype: int64

In [164]:
fsq_y_bar_with_nulls_df['review_count'].describe()

count    219.000000
mean      60.000000
std       78.933952
min        1.000000
25%       12.500000
50%       35.000000
75%       68.000000
max      598.000000
Name: review_count, dtype: float64

In [165]:
# replacing the review_count column within the 25th and 95th quantiles :
fsq_y_bar_with_nulls_df['review_count'] = fsq_y_bar_with_nulls_df.apply(lambda row: 
                                                                        np.random.randint(low= fsq_y_bar_with_nulls_df['review_count'].quantile(0.05), 
                                                                                          high=fsq_y_bar_with_nulls_df['review_count'].quantile(0.95)) if pd.isna(row['review_count']) else row['review_count'], axis=1)
fsq_y_bar_with_nulls_df['review_count'].describe()

count    487.000000
mean      86.291581
std       75.688644
min        1.000000
25%       26.000000
50%       60.000000
75%      142.500000
max      598.000000
Name: review_count, dtype: float64

In [166]:
# Create a new bar_id column that combines the two id columnns (fsq_id and y_id) into one
fsq_y_bar_with_nulls_df['bar_id'] =  fsq_y_bar_with_nulls_df.apply(lambda row: row['y_id'] if pd.isna(row['fsq_id']) else row['fsq_id'], axis=1)
fsq_y_bar_with_nulls_df.drop(columns=['fsq_id', 'y_id'], inplace=True)

# check if it worked. It should be zeros all the way
fsq_y_bar_with_nulls_df.isnull().sum()


bar_name        0
station_id      0
popularity      0
review_count    0
price           0
total_counts    0
rating          0
bar_id          0
dtype: int64

In [167]:
#Renaming to help reduce naming confusion:
clean_fsq_y_bar_df = fsq_y_bar_with_nulls_df #since the with_nulls_df now has no nulls

clean_fsq_y_bar_df.describe()




popularity  review_count       price  total_counts      rating
count  487.000000    487.000000  487.000000    487.000000  487.000000
mean     0.718477     86.291581    1.977413      6.236140    7.722382
std      0.312244     75.688644    0.592740      5.562189    1.033399
min      0.003384      1.000000    1.000000      1.000000    2.000000
25%      0.476300     26.000000    2.000000      2.000000    7.000000
50%      0.899619     60.000000    2.000000      4.000000    8.000000
75%      0.965410    142.500000    2.000000      8.000000    8.200000
max      0.998544    598.000000    4.000000     43.000000   10.000000

In [168]:
# Saving it as a CSV:
clean_fsq_y_bar_df.to_csv('/Users/zachrosenthal/Desktop/Data Science/PYTHON/KATAS/Jupyter Notebooks/Statistical-Modelling-Project/data/clean_fsq_y_bar_df.csv')

In [169]:
# Reading the CSV to test:
clean_fsq_y_bar_df = pd.read_csv('/Users/zachrosenthal/Desktop/Data Science/PYTHON/KATAS/Jupyter Notebooks/Statistical-Modelling-Project/data/clean_fsq_y_bar_df.csv', index_col=0)

clean_fsq_y_bar_df.head()

bar_name                        station_id  \
0                                 Le 4e Mur  36c6491aa1b52e5ef7005f984738de27   
1                             Le Saint Bock  36c6491aa1b52e5ef7005f984738de27   
2  Brasserie Artisanale l'Amère a Boire Inc  36c6491aa1b52e5ef7005f984738de27   
3                                     Cadet  36c6491aa1b52e5ef7005f984738de27   
4                           Pub l'Ile Noire  36c6491aa1b52e5ef7005f984738de27   

   popularity  review_count  price  total_counts  rating  \
0    0.966045          99.0    3.0          11.0     9.0   
1    0.980111          36.0    2.0           9.0     8.3   
2    0.960864         131.0    2.0           8.0     8.1   
3    0.981131          13.0    2.0          23.0     9.5   
4    0.975758         129.0    2.0           6.0     7.8   

                     bar_id  
0  55aedbd9498eb6db5d095733  
1  4b0c1ff8f964a520603723e3  
2  4ad4c06bf964a5203ef920e3  
3  56ec4544498e27e294330c2b  
4  4c1009053ce120a19f6d071c

In [170]:
# Reorganizing the dataframe so that it has all the numerical values on the same side:
clean_fsq_y_bar_df = clean_fsq_y_bar_df[['bar_id', 'bar_name', 'station_id', 'popularity', 'rating', 'price', 'review_count', 'total_counts']]
clean_fsq_y_bar_df

bar_id                                  bar_name  \
0    55aedbd9498eb6db5d095733                                 Le 4e Mur   
1    4b0c1ff8f964a520603723e3                             Le Saint Bock   
2    4ad4c06bf964a5203ef920e3  Brasserie Artisanale l'Amère a Boire Inc   
3    56ec4544498e27e294330c2b                                     Cadet   
4    4c1009053ce120a19f6d071c                           Pub l'Ile Noire   
..                        ...                                       ...   
518    uaf_yrwWo35aWM0_qJCgrw                                  Pincette   
519    ZntrjpC6xAPHPVIYc_TUmA                        John Michael's Pub   
520    jsaZ2JDgCMl2uk2cRmVPjg                               Pub St-Paul   
524    YMJ7E1AcjDF5kN7hVkWd_w                                    Vargas   
525    EyCO8HurmdWYbVCdxPg-Ig                       L'Gros Luxe Plateau   

                           station_id  popularity  rating  price  \
0    36c6491aa1b52e5ef7005f984738de27    0.966045     9.0    3.0   
1    36c6491aa1b52e5ef7005f984738de27    0.980111     8.3    2.0   
2    36c6491aa1b52e5ef7005f984738de27    0.960864     8.1    2.0   
3    36c6491aa1b52e5ef7005f984738de27    0.981131     9.5    2.0   
4    36c6491aa1b52e5ef7005f984738de27    0.975758     7.8    2.0   
..                                ...         ...     ...    ...   
518  3cc8af25d891cb13e3301e4a612d0383    0.331555     9.0    3.0   
519  3cc8af25d891cb13e3301e4a612d0383    0.780364     6.0    2.0   
520  3cc8af25d891cb13e3301e4a612d0383    0.649107     5.0    2.0   
524  7214c61eb7f8d875d6db8db935ad16ea    0.551427     7.0    3.0   
525  15f3a0b338dba26f99a8df6fab3fdbc7    0.571644     7.0    2.0   

     review_count  total_counts  
0            99.0          11.0  
1            36.0           9.0  
2           131.0           8.0  
3            13.0          23.0  
4           129.0           6.0  
..            ...           ...  
518          38.0           5.0  
519          48.0           4.0  
520          51.0           3.0  
524          95.0           1.0  
525          42.0           1.0  

[487 rows x 8 columns]

In [171]:
#Saving the reorganized dataframe to replace the previous one:
clean_fsq_y_bar_df.to_csv('/Users/zachrosenthal/Desktop/Data Science/PYTHON/KATAS/Jupyter Notebooks/Statistical-Modelling-Project/data/clean_fsq_y_bar_df.csv')

clean_bar_df = pd.read_csv('/Users/zachrosenthal/Desktop/Data Science/PYTHON/KATAS/Jupyter Notebooks/Statistical-Modelling-Project/data/clean_fsq_y_bar_df.csv', index_col=0)

clean_bar_df.head()

bar_id                                  bar_name  \
0  55aedbd9498eb6db5d095733                                 Le 4e Mur   
1  4b0c1ff8f964a520603723e3                             Le Saint Bock   
2  4ad4c06bf964a5203ef920e3  Brasserie Artisanale l'Amère a Boire Inc   
3  56ec4544498e27e294330c2b                                     Cadet   
4  4c1009053ce120a19f6d071c                           Pub l'Ile Noire   

                         station_id  popularity  rating  price  review_count  \
0  36c6491aa1b52e5ef7005f984738de27    0.966045     9.0    3.0          99.0   
1  36c6491aa1b52e5ef7005f984738de27    0.980111     8.3    2.0          36.0   
2  36c6491aa1b52e5ef7005f984738de27    0.960864     8.1    2.0         131.0   
3  36c6491aa1b52e5ef7005f984738de27    0.981131     9.5    2.0          13.0   
4  36c6491aa1b52e5ef7005f984738de27    0.975758     7.8    2.0         129.0   

   total_counts  
0          11.0  
1           9.0  
2           8.0  
3          23.0  
4           6.0